File: exam-system-java/entity/src/main/java/com/yyjy/exam/entity/question/entity/Categories.java
package com.yyjy.exam.entity.question.entity;

import org.babyfish.jimmer.Formula;
import org.babyfish.jimmer.sql.*;
import org.jetbrains.annotations.Nullable;

import java.time.LocalDateTime;
import java.util.List;

@Entity
public interface Categories {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    long id();

    @Key
    String name();

    @ManyToOne
    Categories parent();
    
    @IdView
    long parentId();

    int sort();

    @OneToMany(mappedBy = "category")
    List<Questions> questions();

    @OneToMany(mappedBy = "parent")
    List<Categories> children();

    //@Formula(sql = "select count(*) from questions q where q.category_id = ${this}.id")
    //int questionCount();
    @Formula(dependencies = {"questions"})
    default int questionCount() {
        return questions().size();
    }

    @Nullable
    LocalDateTime createTime();

    @Nullable
    LocalDateTime updateTime();

    @Column(name = "deleted_at")
    @LogicalDeleted("now")
    @Nullable
    LocalDateTime delFlag();
}

File: exam-system-server/src/main/kotlin/com/yyjy/models/entity/Categories.kt
package com.yyjy.models.entity

import org.babyfish.jimmer.Formula
import org.babyfish.jimmer.sql.*
import java.time.LocalDateTime

/**
 * 题目分类表（分类体系）
 */
@Entity
interface Categories {

    @Id
    @GeneratedValue(
        strategy = GenerationType.IDENTITY
    )
    val id: Long

    @Key
    val name: String

    @ManyToOne
    val parent: Categories

    @IdView
    val parentId: Long

    val sort: Int

    @OneToMany(mappedBy = "category")
    val questions: List<Questions>

    @OneToMany(mappedBy = "parent")
    val children: List<Categories>

    /**
     * 创建时间
     */
    val createTime: LocalDateTime?

    /**
     * 更新时间
     */
    val updateTime: LocalDateTime?

    @Formula(dependencies = ["questions"])
    val questionCount: Int
        get() = questions.size

    /**
     * 删除时间
     */
    @Column(name = "deleted_at")
    @LogicalDeleted("now")
    val delFlag: LocalDateTime?
}

